examples/99_in_tool_selection/ex_in_tool_selection/main.rb

```diff
@@ -0,0 +1,364 @@
+        # Otherwise keep this preselection.
+        Sketchup.active_model.selection.clear unless valid_selection?
+
+        # The tool selection is tracked separately from the SketchUp selection.
+        # When we are in the selection stage we want to temporarily select
+        # hovered entities for previewing them as selected, but our tool logic
+        # treats the selection as empty until the user clicks.
+        #
+        # Native tools can preview entities as selected without actually
+        # adding them to the selection, but the Ruby API doesn't support this.
+        @selection = Sketchup.active_model.selection.to_a
+
+        # Remember whether tool was activated with preselection or not as this
+        # affect how the tool resets.
+        @preselection = !Sketchup.active_model.selection.empty?
+
+        # The entity being hovered when at the selection stage.
+        @hovered_entity = nil
+
+        @valid_hovered_entity = false
+
+        update_ui
+      end
+
+      def deactivate(view)
+        @inference_lock_helper.unlock
+        view.model.abort_operation if @in_operation
+        view.invalidate
+      end
+
+      def enableVCB?
+        picked_first_point?
+      end
+
+      def resume(view)
+        update_ui
+        view.invalidate
+      end
+
+      def suspend(view)
+        view.invalidate
+      end
+
+      def onCancel(reason, view)
+        reset_tool
+        view.invalidate
+      end
+
+      def onKeyDown(key, _repeat, _flags, view)
+        @inference_lock_helper.on_keydown(key, view, @mouse_ip, @picked_first_ip)
+        pick_mouse_position(view)
+      end
+
+      def onKeyUp(key, _repeat, _flags, view)
+        @inference_lock_helper.on_keyup(key, view)
+        pick_mouse_position(view)
+      end
+
+      def onMouseMove(flags, x, y, view)
+        if @selection.empty?
+          # If the tool selection is empty, try picking the hovered object.
+          try_select_entity(view, x, y)
+        end
+
+        @mouse_position = Geom::Point3d.new(x, y, 0)
+        pick_mouse_position(view)
+      end
+
+      def onLButtonDown(flags, x, y, view)
+        @mouse_down = Geom::Point3d.new(x, y)
+
+        # If tool selection is empty, pick our hovered entity.
+        if @selection.empty?
+          @selection = view.model.selection.to_a
+        end
+
+        return if @selection.empty?
+
+        # For a Move or Rotate like tool, we can select a starting point in the
+        # same mouse click as we select an entity. For Scale or other tools
+        # where you need to select a handle, this logic is typically slightly
+        # different, and separate click is needed.
+
+        if !picked_first_point?
+          start_move(view)
+        else
+          end_move(view)
+        end
+
+        update_ui
+        view.invalidate
+      end
+
+      def onLButtonUp(flags, x, y, view)
+        if @mouse_down.distance([x, y]) > DRAG_THRESHOLD
```

Comment:
No need to reset anything if there wasn't a drag?

Reply:
You could set `@mouse_down` to nil or ORIGIN, but I don't know if it's needed.

Reply:
Ok, just checking as I often find tool state needs a few members to be reset.